mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala

```diff
@@ -0,0 +1,466 @@
+
+  /**
+   * Sets the maximum number of iterations.
+   * Default is 100.
+   * @group setParam
+   */
+  @Since("2.1.0")
+  def setMaxIter(value: Int): this.type = set(maxIter, value)
+  setDefault(maxIter -> 100)
+
+  /**
+   * Sets the convergence tolerance of iterations.
+   * Smaller value will lead to higher accuracy with the cost of more iterations.
+   * Default is 1E-6.
+   * @group setParam
+   */
+  @Since("2.1.0")
+  def setTol(value: Double): this.type = set(tol, value)
+  setDefault(tol -> 1E-6)
+
+  /**
+   * Whether to over-/under-sample training instances according to the given weights in weightCol.
+   * If not set or empty, all instances are treated equally (weight 1.0).
+   * Default is not set, so all instances have weight one.
+   * @group setParam
+   */
+  @Since("2.1.0")
+  def setWeightCol(value: String): this.type = set(weightCol, value)
+
+  override protected def train(dataset: Dataset[_]): RobustRegressionModel = {
+    val numFeatures = dataset.select(col($(featuresCol))).first().getAs[Vector](0).size
+    val w = if (!isDefined(weightCol) || $(weightCol).isEmpty) lit(1.0) else col($(weightCol))
+
+    val instances: RDD[Instance] =
+      dataset.select(col($(labelCol)), w, col($(featuresCol))).rdd.map {
+        case Row(label: Double, weight: Double, features: Vector) =>
+          Instance(label, weight, features)
+      }
+
+    val handlePersistence = dataset.rdd.getStorageLevel == StorageLevel.NONE
+    if (handlePersistence) instances.persist(StorageLevel.MEMORY_AND_DISK)
+
+    val featuresSummarizer = {
+      val seqOp = (c: MultivariateOnlineSummarizer, v: Instance) => c.add(v.features, v.weight)
+      val combOp = (c1: MultivariateOnlineSummarizer, c2: MultivariateOnlineSummarizer) => {
+        c1.merge(c2)
+      }
+      instances.treeAggregate(new MultivariateOnlineSummarizer)(seqOp, combOp)
+    }
+
+    val featuresStd = featuresSummarizer.variance.toArray.map(math.sqrt)
+    val regParamL2 = $(regParam)
+
+    val costFun = new HuberCostFun(instances, $(fitIntercept), $(standardization),
+      featuresStd, regParamL2, $(m))
+
+//    val lowerBounds = BDV[Double](Array.fill(numFeatures + 2)(Double.MinValue))
+//    lowerBounds(0) = 1E-20
+//    val upperBounds = BDV[Double](Array.fill(numFeatures + 2)(Double.MaxValue))
+
+//    val optimizer = new BreezeLBFGSB(lowerBounds, upperBounds, $(maxIter), 10, $(tol))
+    val optimizer = new BreezeLBFGS[BDV[Double]]($(maxIter), 10, $(tol))
+
+    val initialArray = Array.fill(numFeatures + 2)(1E-20)
+    if (!$(fitIntercept)) initialArray(1) = 0.0
+    val initialParameters = Vectors.dense(initialArray)
+
+    val states = optimizer.iterations(new CachedDiffFunction(costFun),
+      initialParameters.asBreeze.toDenseVector)
+
+    val parameters = {
+      val arrayBuilder = mutable.ArrayBuilder.make[Double]
+      var state: optimizer.State = null
+      while (states.hasNext && !(state != null && state.searchFailed)) {
```

Comment:
Could you `states.foreach` or similar with `filter`? Double negation begs for a simpler solution.